File: lib_baseapp/src/main/java/com/liuxd/baseapplibrary/base/BaseAdapter.java
package com.liuxd.baseapplibrary.base;

import androidx.appcompat.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

/**
 * Created by Liuxd on 2016/11/8 14:55.
 * <p>
 * RecyclerView 通用的适配器
 * </p>
 * 通过convert（）抽象方法将holder及数据公布出去，以及基本的点击、长按事件监听处理
 */

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<BaseViewHolder> {
    private List<T> mDatas;
    private int mLayoutId;
    private OnItemClickListener mOnItemClickListener;
    private OnItemLongClickListener mOnItemLongClickListener;

    /**
     * @param datas    数据源
     * @param layoutId 布局id
     */
    public BaseAdapter(List<T> datas, int layoutId) {
        this.mDatas = datas;
        this.mLayoutId = layoutId;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(mLayoutId, parent, false);
        return new BaseViewHolder(parent.getContext(), itemView);
    }

    @Override
    public void onBindViewHolder(final BaseViewHolder holder, int position) {
        convert(holder, mDatas.get(position));
        //设置点击事件
        if (mOnItemClickListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    mOnItemClickListener.onItemClickListener(v, holder.getAdapterPosition());
                }

            });
        }
        //设置长按事件
        if (mOnItemLongClickListener == null) {
            holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    return mOnItemLongClickListener.onItemLongClickListener(v, holder.getAdapterPosition());
                }
            });
        }

    }

    @Override
    public int getItemCount() {
        return mDatas != null ? mDatas.size() : 0;
    }

    public void setDatas(List<T> datas, boolean loadMore) {
        if (loadMore) {
            this.mDatas.addAll(datas);
        } else {
            this.mDatas = datas;
        }
        notifyDataSetChanged();
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.mOnItemClickListener = onItemClickListener;
    }

    public void setOnItemLongClickListener(OnItemLongClickListener onItemLongClickListener) {
        this.mOnItemLongClickListener = onItemLongClickListener;
    }

    public interface OnItemClickListener {
        void onItemClickListener(View v, int position);
    }

    public interface OnItemLongClickListener {
        boolean onItemLongClickListener(View v, int position);
    }

    public abstract void convert(BaseViewHolder holder, T data);
}

File: app/src/main/java/com/asura/android_study/ui/base/BaseAdapter.kt
package com.asura.android_study.ui.base

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

/**
 * Created by Liuxd on 2016/11/8 14:55.
 *
 *
 * RecyclerView 通用的适配器
 *
 * 通过convert（）抽象方法将holder及数据公布出去，以及基本的点击、长按事件监听处理
 * @param datas    数据源
 * @param layoutId 布局id
 */
abstract class BaseAdapter<T> constructor(
    private var mDatas: MutableList<T>?,
    private val mLayoutId: Int
) : RecyclerView.Adapter<BaseViewHolder>() {
    private var mOnItemClickListener: OnItemClickListener? = null
    private var mOnItemLongClickListener: OnItemLongClickListener? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(mLayoutId, parent, false)
        return BaseViewHolder(parent.context, itemView)
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        convert(holder, mDatas!![position])
        //设置点击事件
        if (mOnItemClickListener != null) {
            holder.itemView.setOnClickListener { v -> mOnItemClickListener!!.onItemClickListener(v, holder.adapterPosition) }
        }
        //设置长按事件
        if (mOnItemLongClickListener == null) {
            holder.itemView.setOnLongClickListener { v -> mOnItemLongClickListener!!.onItemLongClickListener(v, holder.adapterPosition) }
        }
    }

    override fun getItemCount(): Int {
        return if (mDatas != null) mDatas!!.size else 0
    }

    fun setDatas(datas: MutableList<T>?, loadMore: Boolean) {
        if (loadMore) {
            mDatas!!.addAll(datas!!)
        } else {
            mDatas = datas
        }
        notifyDataSetChanged()
    }

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener?) {
        mOnItemClickListener = onItemClickListener
    }

    fun setOnItemLongClickListener(onItemLongClickListener: OnItemLongClickListener?) {
        mOnItemLongClickListener = onItemLongClickListener
    }

    interface OnItemClickListener {
        fun onItemClickListener(v: View?, position: Int)
    }

    interface OnItemLongClickListener {
        fun onItemLongClickListener(v: View?, position: Int): Boolean
    }

    abstract fun convert(holder: BaseViewHolder, data: T)
}